sdk/src/slot_history.rs

```diff
@@ -31,11 +46,37 @@ pub enum Check {
 
 impl SlotHistory {
     pub fn add(&mut self, slot: Slot) {
-        for skipped in self.next_slot..slot {
-            self.bits.set(skipped % MAX_ENTRIES, false);
+        if slot > self.next_slot && slot - self.next_slot >= MAX_ENTRIES {
+            // Wrapped past current history,
+            // clear entire bitvec.
+            let full_blocks = (MAX_ENTRIES as usize) / 64;
+            for i in 0..full_blocks {
+                self.bits.set_block(i, 0);
+            }
+            self.next_slot = slot + 1;
+        } else if slot >= self.next_slot {
+            // Current and next pointers are within
+            // the same MAX_ENTRIES slots
+            let start = self.next_slot % MAX_ENTRIES;
+            let end = slot % MAX_ENTRIES;
+            if end >= start {
+                // Straightforward clear between old head and new head
+                for skipped in start..end {
+                    self.bits.set(skipped, false);
+                }
+            } else {
```

Comment:
nits: I don't see much of merits of branching between this and [here](https://github.com/solana-labs/solana/pull/10175/files#diff-b6bc75c0bdcd2062ec00c3da3d974053R62).

I think [the old code](https://github.com/solana-labs/solana/pull/10175/files#diff-b6bc75c0bdcd2062ec00c3da3d974053L34) could just be indented here.

Not like [the preceding `set_block`](https://github.com/solana-labs/solana/pull/10175/files#diff-b6bc75c0bdcd2062ec00c3da3d974053R53), the same number of `self.bit.set` is still called here.

Yeah, we will reduce the the number of `%` operation, but we got more code instead.

Anyway the current new code works as intended!

Reply:
Yea. I considered that as well. The old code is attractively small. I did want to reduce the inner-loop operations. As you said for this case, the mod is reduced to shift, but you also might want to have a non-power-of-2 number of entries where that will not be the case. Let me run a quick bench, maybe that will resolve this.